### ISSUE TITLE
Pre-processing of downloaded submission files

### ISSUE THREAD

**Issue by lgpage** [jupyter/nbgrader#511]
I would like to propose adding a new feature of being able to pre-process submission files.

For example, institutions using BlackBoard can download assignment submission files in a `.zip` archive with the following filename format (give or take some extra garbage in the filename):
`{couse_id}{assignment_id}{timestamp}.zip`

Each of the files in this archive will have the following filename format (give or take some extra garbage in the filename):
`{assignment_id}{student_id}{timestamp}{upload_filename}.{ext}`

I propose allowing for the following (optional) directories in the nbgrader directory structure:
`{course_directory}/downloaded/archive/{zipfiles}`
`{course_directory}/downloaded/{assignment_id}/extracted/{raw_files}.ipynb`
`{course_directory}/downloaded/{assignment_id}/invalid/{raw_files}.{ext}`

The pre-processing step will do two things:
1. extract the archive file/s to `{course_directory}/downloaded/{assignment_id}/extracted/`
2. process these extracted files into `{course_directory}/downloaded/{assignment_id}/invalid/` or the submitted path `{course_directory}/submitted/{student_id}/{assignment_id}/`

This needs to be done in a pluggable way (similar to what is mentioned in #233) as filenames will vary from user to user. There may also the possible option of automatically being able to update the students in the database, based on this processing.

**Comment by jhamrick** [comment]
This seems like a reasonable thing to have. I guess this would call for a new command, maybe something along the lines of `nbgrader import` (I am thinking "import" specifically because I've been thinking of having some sort of "export" application related to #233 and also #436, so having it be "import" would give it a nice symmetry).

Definitely it should be pluggable -- I'm thinking maybe we can define a new type of `ImportPlugin` class (or something along those lines) that does all the heavy lifting. We could maintain a few popular ones (e.g. blackboard and canvas) but still allow other people to write their own if they are using something different.

Would you like to take a stab at a PR for this?

**Comment by lgpage** [comment]
A new command would be needed, I was thinking something along the lines of `nbgrader stage`, although `import` does give nice symmetry to `export`

For pluggability, my initial thought was to allow for specific functions in the `nbgrader_config.py` file that allows for the customized behaviour. Wrapping this in a `PluginManager` (of sorts) for default behaviours and/or supported behaviours (e.g for blackboard and canvas) is a good idea :)

With respects to #233, allow something like this in the `nbgrader_config.py` file:
```python
def late_penalty(student_id, score, duedate, extension, timestamp):
    return penalty_score
```

I'll take a stab at a PR for this.
Might take a little while, I've only just started diving into the source code and still trying to wrap my head around it :P

**Comment by jhamrick** [comment]
> I'll take a stab at a PR for this.

Awesome -- let me know if you have any questions about any parts of the code!

**Comment by lgpage** [comment]
closed by #526
